src/dxil/wrapper.rs

```diff
@@ -0,0 +1,480 @@
+    inner: ComPtr<ID3D12FunctionReflection>,
+}
+
+impl D3D12FunctionReflection {
+    fn new(inner: ComPtr<ID3D12FunctionReflection>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_constant_buffer_by_index(
+        &self,
+        buffer_index: u32,
+    ) -> D3D12ShaderReflectionConstantBuffer {
+        unsafe {
+            D3D12ShaderReflectionConstantBuffer::new(
+                self.inner.get_constant_buffer_by_index(buffer_index),
+            )
+        }
+    }
+
+    pub fn get_constant_buffer_by_name(&self, name: &CStr) -> D3D12ShaderReflectionConstantBuffer {
+        unsafe {
+            D3D12ShaderReflectionConstantBuffer::new(
+                self.inner
+                    .get_constant_buffer_by_name(name.as_ptr() as *const i8),
+            )
+        }
+    }
+
+    pub fn get_desc(&self) -> Result<d3d12shader::D3D12_FUNCTION_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_FUNCTION_DESC::default();
+        return_hr_wrapped!(unsafe { self.inner.get_desc(&mut desc as *mut _) }, desc);
+    }
+
+    pub fn get_function_parameter(&self, parameter_index: i32) -> D3D12FunctionParameterReflection {
+        unsafe {
+            D3D12FunctionParameterReflection::new(
+                self.inner.get_function_parameter(parameter_index),
+            )
+        }
+    }
+
+    pub fn get_resource_binding_desc(
+        &self,
+        resource_index: u32,
+    ) -> Result<d3d12shader::D3D12_SHADER_INPUT_BIND_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_SHADER_INPUT_BIND_DESC::default();
+        return_hr_wrapped!(
+            unsafe {
+                self.inner
+                    .get_resource_binding_desc(resource_index, &mut desc as *mut _)
+            },
+            desc
+        );
+    }
+
+    pub fn get_resource_binding_desc_by_name(
+        &self,
+        name: &CStr,
+    ) -> Result<d3d12shader::D3D12_SHADER_INPUT_BIND_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_SHADER_INPUT_BIND_DESC::default();
+        return_hr_wrapped!(
+            unsafe {
+                self.inner.get_resource_binding_desc_by_name(
+                    name.as_ptr() as *const i8,
+                    &mut desc as *mut _,
+                )
+            },
+            desc
+        );
+    }
+
+    pub fn get_variable_by_name(&self, name: &CStr) -> D3D12ShaderReflectionVariable {
+        unsafe {
+            D3D12ShaderReflectionVariable::new(
+                self.inner.get_variable_by_name(name.as_ptr() as *const i8),
+            )
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct D3D12ShaderReflectionType {
+    inner: ComPtr<ID3D12ShaderReflectionType>,
+}
+
+impl D3D12ShaderReflectionType {
+    fn new(inner: ComPtr<ID3D12ShaderReflectionType>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_base_class(&self) -> D3D12ShaderReflectionType {
+        unsafe { D3D12ShaderReflectionType::new(self.inner.get_base_class()) }
```

Comment:
The documentation doesn't say anything about `NULL` returns here but given that this is a recursive function (can be repeatedly called on the returned value) there should be an end?

All I can find in DxCompiler source is the method on `CShaderReflectionType` returning null, and on `CInvalidSRType` returning a pointer to a static version on itself. The rest of the implementation might reside elsewhere?

The FFI layer is able to deal with `Option<>` to indicate a nullable value, you might want to wrap the ptr in that and get a natural way to expect something or nothing.
https://anssi-fr.github.io/rust-guide/07_ffi.html#note-2

Be sure to check the other functions returning an interface ptr as well.